docs/source/cpp/compute.rst

```diff
@@ -0,0 +1,419 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. default-domain:: cpp
+.. highlight:: cpp
+.. cpp:namespace:: arrow::compute
+
+=================
+Compute Functions
+=================
+
+.. TODO: describe API and how to invoke compute functions
```

Comment:
```suggestion
The inputs and outputs of compute functions are of type :class:`arrow::Datum`, which are
a discriminated union of several shapes of data, including :class:`arrow::Scalar`,
:class:`arrow::Array`, and :class:`arrow::ChunkedArray`.

Compute functions can be invoked by name using the :func:`arrow::compute::CallFunction`
or via dedicated c++ convenience functions.

.. code-block:: cpp
   std::shared_ptr<arrow::Int32Array> ints = ...;
   std::shared_ptr<arrow::Int32Scalar> increment = ...;
   
   ARROW_ASSIGN_OR_RAISE(arrow::Datum incremented, arrow::compute::CallFunction("add_checked",
       {arrow::Datum(ints), arrow::Datum(increment)}));

   arrow::ArithmeticOptions options;
   options.check_overflow = true;
   ARROW_ASSIGN_OR_RAISE(arrow::Datum equivalent, arrow::compute::Add(
       {arrow::Datum(ints), arrow::Datum(increment)}, arithmetic_options));
   
```

Reply:
Note: it isn't necessary to use `Datum(...)` explicitly on most cases because of implicit conversions

Reply:
This is true, the explicit typing was for clarity of exposition